protocol/testing/e2e/funding/funding_e2e_test.go

```diff
@@ -0,0 +1,247 @@
+		},
+		Side: clobtypes.Order_SIDE_SELL,
+		GoodTilOneof: &clobtypes.Order_GoodTilBlockTime{
+			GoodTilBlockTime: uint32(GenesisTime.Add(24 * time.Hour).Unix()),
+		},
+	}
+	OrderTemplate_Bob_Num0_Id1_Clob0_Sell_LongTerm = clobtypes.Order{
+		OrderId: clobtypes.OrderId{
+			SubaccountId: constants.Bob_Num0,
+			ClientId:     1,
+			OrderFlags:   clobtypes.OrderIdFlags_LongTerm,
+			ClobPairId:   0,
+		},
+		Side: clobtypes.Order_SIDE_SELL,
+		GoodTilOneof: &clobtypes.Order_GoodTilBlockTime{
+			GoodTilBlockTime: uint32(GenesisTime.Add(24 * time.Hour).Unix()),
+		},
+	}
+	// Genesis time of the chain
+	GenesisTime                               = time.Unix(1690000000, 0)
+	FirstFundingSampleTick                    = time.Unix(1690000050, 0)
+	FirstFundingTick                          = time.Unix(1690002000, 0)
+	LastFundingSampleOfSecondFundingTickEpoch = time.Unix(1690005570, 0)
+	SecondFundingTick                         = time.Unix(
+		1690002000+int64(epochstypes.FundingTickEpochDuration),
+		0,
+	)
+)
+
+func TestFunding(t *testing.T) {
+	tests := map[string]struct {
+		testHumanOrders   []TestHumanOrder
+		initialIndexPrice map[uint32]string
+		// index price to be used in premium calculation
+		indexPriceForPremium map[uint32]string
+		// oracle price for funding index calculation
+		oracelPriceForFundingIndex map[uint32]string
+		expectedFundingPremium     int32
+		expectedFundingIndex       int64
+	}{
+		"Test funding": {
+			testHumanOrders: []TestHumanOrder{
+				// Unmatched orders to generate funding premiums.
+				{
+					Order:      OrderTemplate_Bob_Num0_Id0_Clob0_Sell_LongTerm,
+					HumanPrice: "28005",
+					HumanSize:  "2",
+				},
+				{
+					Order:      OrderTemplate_Alice_Num0_Id0_Clob0_Buy_LongTerm,
+					HumanPrice: "28000",
+					HumanSize:  "2",
+				},
+				// Matched orders to set up Alice and Bob's positions.
+				{
+					Order:      OrderTemplate_Bob_Num0_Id1_Clob0_Sell_LongTerm,
+					HumanPrice: "28003",
+					HumanSize:  "1",
+				},
+				{
+					Order:      OrderTemplate_Alice_Num0_Id1_Clob0_Buy_LongTerm,
+					HumanPrice: "28003",
+					HumanSize:  "1",
+				},
+			},
+			initialIndexPrice: map[uint32]string{
+				0: "28002",
+			},
+			indexPriceForPremium: map[uint32]string{
+				0: "27960",
+			},
+			oracelPriceForFundingIndex: map[uint32]string{
+				0: "27000",
+			},
+			expectedFundingPremium: 1430, // 28_000 / 27_960 - 1 ~= 0.001430
+			// 1430 / 8 * 27000 * 10^(btc_atomic_resolution - quote_atomic_resolution) ~= 483
+			expectedFundingIndex: 483,
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			tApp := testapp.NewTestAppBuilder(t).WithGenesisDocFn(func() (genesis types.GenesisDoc) {
+				genesis = testapp.DefaultGenesis()
+				genesis.GenesisTime = GenesisTime
+				return genesis
+			}).Build()
+			ctx := tApp.InitChain()
+
+			// Place orders on the book.
+			for _, testHumanOrder := range tc.testHumanOrders {
+				order := clobtest.MustMakeOrderFromHumanInput(
```

Comment:
It would be nice if we could just use [MustScaleOrder](https://github.com/dydxprotocol/v4-chain/blob/a612627dc7bea419a23bbe2307636818a36b0285/protocol/x/clob/e2e/app_test.go#L681) but understand that it may not have the granularity that you need for these tests.